core/templates/pages/diagnostic-test-player-page/diagnostic-test.model.ts

```diff
@@ -0,0 +1,256 @@
+  getFailedTopicIds: () => string[];
+  getCurrentTopicId: () => string;
+  setCurrentTopicId: () => void;
+}
+
+export class DiagnosticTestModelData {
+  _totalNumberOfAttemptedQuestions: number;
+  _currentTopicId: string;
+  _eligibleTopicIds: string[];
+  _skippedTopicIds: string[];
+  _passedTopicIds: string[];
+  _failedTopicIds: string[];
+  _topicIdToPrerequisiteTopicIds: TopicIdToRelatedTopicIds;
+  _topicIdToAncestorTopicIds: TopicIdToRelatedTopicIds;
+  _topicIdToSuccessorTopicIds: TopicIdToRelatedTopicIds;
+
+  constructor(topicIdToPrerequisiteTopicIds: TopicIdToRelatedTopicIds) {
+    this._topicIdToPrerequisiteTopicIds = topicIdToPrerequisiteTopicIds;
+    this._totalNumberOfAttemptedQuestions = 0;
+    this._eligibleTopicIds = Object.keys(this._topicIdToPrerequisiteTopicIds);
+    this._skippedTopicIds = [];
+    this._passedTopicIds = [];
+    this._failedTopicIds = [];
+    this._currentTopicId = '';
+    this._topicIdToAncestorTopicIds = {};
+    this._topicIdToSuccessorTopicIds = {};
+    this.setTopicIdToAncestorTopicIds();
+    this.setTopicIdToSuccessorTopicIds();
+  }
+
+  getEligibleTopicIds(): string[] {
+    return this._eligibleTopicIds;
+  }
+
+  getFailedTopicIds(): string[] {
+    return this._failedTopicIds;
+  }
+
+  getPassedTopicIds(): string[] {
+    return this._passedTopicIds;
+  }
+
+  getCurrentTopicId(): string {
+    return this._currentTopicId;
+  }
+
+  getAncestorsTopicIds(topicId: string): string[] {
+    return this._topicIdToAncestorTopicIds[topicId];
+  }
+
+  getSkippedTopicIds(): string[] {
+    return this._skippedTopicIds;
+  }
+
+  getTopicIdToAncestorTopicIds(): TopicIdToRelatedTopicIds {
+    return this._topicIdToAncestorTopicIds;
+  }
+
+  getTopicIdToSuccessorTopicIds(): TopicIdToRelatedTopicIds {
+    return this._topicIdToSuccessorTopicIds;
+  }
+
+  getSuccessorTopicIds(topicId: string): string[] {
+    return this._topicIdToSuccessorTopicIds[topicId];
+  }
+
+  incrementNumberOfAttemptedQuestions(incrementByValue: number): void {
+    this._totalNumberOfAttemptedQuestions += incrementByValue;
+  }
+
+  getNumberOfAttemptedQuestions(): number {
+    return this._totalNumberOfAttemptedQuestions;
+  }
+
+  setTopicIdToAncestorTopicIds(): void {
+    for (let topicId in this._topicIdToPrerequisiteTopicIds) {
+      let ancestors: string[] = [];
+      let prerequisites: string[] = cloneDeep(
+        this._topicIdToPrerequisiteTopicIds[topicId]);
+
+      while (prerequisites.length > 0) {
+        let len = prerequisites.length;
+        let lastTopicId = prerequisites[len - 1];
+        prerequisites.splice(len - 1, 1);
+        if (ancestors.indexOf(lastTopicId) === -1) {
+          ancestors.push(lastTopicId);
+        }
+
+        prerequisites = prerequisites.concat(
+          this._topicIdToPrerequisiteTopicIds[lastTopicId]);
+      }
+      this._topicIdToAncestorTopicIds[topicId] = ancestors;
+    }
+  }
+
+  setTopicIdToSuccessorTopicIds(): void {
```

Comment:
Ditto re naming, and I think many of the comments for the previous method might apply here too.

Reply:
Renamed